pkg/controller/perconaservermongodb/users.go

```diff
@@ -0,0 +1,241 @@
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const internalPrefix = "internal-"
+
+type sysUser struct {
+	Name string `yaml:"username"`
+	Pass string `yaml:"password"`
+}
+
+func (r *ReconcilePerconaServerMongoDB) reconcileUsers(cr *api.PerconaServerMongoDB) error {
+	sysUsersSecretObj := corev1.Secret{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      cr.Spec.Secrets.Users,
+		},
+		&sysUsersSecretObj,
+	)
+	if err != nil && k8serrors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return errors.Wrapf(err, "get sys users secret '%s'", cr.Spec.Secrets.Users)
+	}
+
+	secretName := internalPrefix + cr.Name + "-users"
+	internalSysSecretObj := corev1.Secret{}
+
+	err = r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      secretName,
+		},
+		&internalSysSecretObj,
+	)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return errors.Wrap(err, "get internal sys users secret")
+	}
+
+	if k8serrors.IsNotFound(err) {
+		internalSysUsersSecret := sysUsersSecretObj.DeepCopy()
+		internalSysUsersSecret.ObjectMeta = metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: cr.Namespace,
+		}
+		err = r.client.Create(context.TODO(), internalSysUsersSecret)
+		if err != nil {
+			return errors.Wrap(err, "create internal sys users secret")
+		}
+		return nil
+	}
+
+	// we do this check after work with secret objects because in case of upgrade cluster we need to be sure that internal secret exist
+	if cr.Status.State != api.AppStateReady {
+		return nil
+	}
+
+	newSysData, err := json.Marshal(sysUsersSecretObj.Data)
+	if err != nil {
+		return errors.Wrap(err, "marshal sys secret data")
+	}
+	newSecretDataHash := sha256Hash(newSysData)
+	dataChanged, err := sysUsersSecretDataChanged(newSecretDataHash, &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "check sys users data changes")
+	}
+
+	if !dataChanged {
+		return nil
+	}
+
+	restartSfs, err := r.manageSysUsers(cr, &sysUsersSecretObj, &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "manage sys users")
+	}
+
+	internalSysSecretObj.Data = sysUsersSecretObj.Data
+	err = r.client.Update(context.TODO(), &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "update internal sys users secret")
+	}
+
+	if restartSfs {
+		sfsTemplateAnnotations["last-applied-secret"] = newSecretDataHash
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaServerMongoDB) manageSysUsers(cr *api.PerconaServerMongoDB, sysUsersSecretObj, internalSysSecretObj *corev1.Secret) (bool, error) {
+	var (
+		sysUsers  []sysUser
+		userAdmin *sysUser
+	)
+	restartSfs := false
+	userAdminUserExist := false
+	userAdminPasswordExist := false
+	for key := range sysUsersSecretObj.Data {
+		switch key {
+		case envMongoDBUserAdminPassword:
+			userAdminPasswordExist = true
+		case envMongoDBUserAdminUser:
+			userAdminUserExist = true
+		}
```

Comment:
you have a `userAdmin ` which is gonna be nil if there is no admin user in the secret, innit? so you don't have to have these vars to check for adminUser presence